pkg/controller/agent/fleet.go

```diff
@@ -0,0 +1,307 @@
+	return response.Item, err
+}
+
+func (f fleetAPI) getEnrollmentAPIKey(ctx context.Context, keyID string) (EnrollmentAPIKey, error) {
+	var response EnrollmentAPIKeyResult
+	err := f.request(ctx, http.MethodGet, fmt.Sprintf("%s/%s", f.enrollmentAPIKeyPath(), keyID), nil, &response)
+	return response.Item, err
+}
+
+func (f fleetAPI) deleteEnrollmentAPIKey(ctx context.Context, keyID string) error {
+	return f.request(ctx, http.MethodDelete, fmt.Sprintf("%s/%s", f.enrollmentAPIKeyPath(), keyID), nil, nil)
+}
+
+func (f fleetAPI) findAgentPolicy(ctx context.Context, filter func(policy Policy) bool) (Policy, error) {
+	page := 1
+	for {
+		var list PolicyList
+		err := f.request(ctx, http.MethodGet, fmt.Sprintf("agent_policies?perPage=20&page=%d", page), nil, &list)
+		if err != nil {
+			return Policy{}, err
+		}
+		if len(list.Items) == 0 {
+			break
+		}
+		for _, p := range list.Items {
+			if filter(p) {
+				return p, nil
+			}
+		}
+		page++
+	}
+	return Policy{}, errors.New("no matching agent policy found")
+}
+
+func (f fleetAPI) defaultFleetServerPolicyID(ctx context.Context) (string, error) {
+	policy, err := f.findAgentPolicy(ctx, func(policy Policy) bool {
+		return policy.IsDefaultFleetServer && policy.Status == "active"
+	})
+	if err != nil {
+		return "", err
+	}
+	return policy.ID, nil
+}
+
+func (f fleetAPI) defaultAgentPolicyID(ctx context.Context) (string, error) {
+	policy, err := f.findAgentPolicy(ctx, func(policy Policy) bool {
+		return policy.IsDefault && policy.Status == "active"
+	})
+	if err != nil {
+		return "", err
+	}
+	return policy.ID, nil
+}
+
+func (f fleetAPI) setupFleet(ctx context.Context) error {
+	return f.request(ctx, http.MethodPost, "setup", nil, nil)
+}
+
+func maybeReconcileFleetEnrollment(params Params, result *reconciler.Results) string {
+	if !params.Agent.Spec.KibanaRef.IsDefined() {
+		return ""
+	}
+
+	reachable, err := isKibanaReachable(params.Context, params.Client, params.Agent.Spec.KibanaRef.WithDefaultNamespace(params.Agent.Namespace).NamespacedName())
+	if err != nil {
+		result.WithError(err)
+		return ""
+	}
+	if !reachable {
+		result.WithResult(reconcile.Result{Requeue: true})
+		return ""
+	}
+
+	kbConnectionSettings, err := extractClientConnectionSettings(params.Context, params.Agent, params.Client, commonv1.KibanaAssociationType)
+	if err != nil {
+		result.WithError(err)
+		return ""
+	}
+
+	token, err := reconcileEnrollmentToken(
+		params.Context, params.Agent, params.Client,
+		newFleetAPI(
+			params.OperatorParams.Dialer,
+			kbConnectionSettings,
+			params.Logger()),
+	)
+	result.WithError(err)
+	return token
+}
+
+func isKibanaReachable(ctx context.Context, client k8s.Client, kibanaNSN types.NamespacedName) (bool, error) {
+	var kb v1.Kibana
+	err := client.Get(ctx, kibanaNSN, &kb)
+	if err != nil {
+		return false, err
+	}
+	if kb.Status.Health != commonv1.GreenHealth {
+		return false, nil // requeue
+	}
+	return true, nil
```

Comment:
I realise this does not work well with external KibanaRefs. I will look into making an actuall HTTP request instead here.

Reply:
It turns out this is a rabbit hole. Making external refs work is more complicated. Other things required: communication between Fleet Server and Elasticsearch needs to use a service token otherwise Agent will try to talk to Kibana to generate one. Also I ran into some weird connectivity issues where Agent seemed to connect to an IP directly without host header and failed to connect through the ESS proxy. Still looking into that, but I might punt the external ref support to another PR if it makes sense at all.